### ISSUE TITLE
Race condition seen when there are multiple readers and writers

### ISSUE THREAD

**Issue by vanti** [ARMmbed/mbedtls#564]
When there are multiple readers and/or writers using the same mbedtls_ssl_context (which is possible in the mbed-client-linux use-case since it has a background data listen thread that continuously performs read operations, and the application thread can independently send 'update registration' or other requests), mutual exclusion should be used to protect the context. For example, this code in mbedtls_ssl_flush_output(), which is called by both mbedtls_ssl_write_record() and mbedtls_ssl_read(), can result in ssl->out_left being corrupted and becoming a huge number (since it is an unsigned int) when reads and writes are performed concurrently:

while( ssl->out_left > 0 )
{
         ...

         ret = ssl->f_send( ssl->p_bio, buf, ssl->out_left );

         MBEDTLS_SSL_DEBUG_RET( 2, "ssl->f_send", ret );

         if( ret <= 0 )
             return( ret );

         ssl->out_left -= ret;
}

**Comment by pjbakker** [comment]
In mbed TLS thread-safe functions are marked as such.

Contexts are only intended to be used from a single thread. If you want to access a context in a multi-threaded way, you'll have to provide the guarding yourself.

See: https://tls.mbed.org/kb/development/thread-safety-and-multi-threading

**Comment by vanti** [comment]
Thank you for the info. I'll bring this up on the mbed client's forum, since they are the users in this case.

**Comment by awakecoding** [comment]
@vanti I am experiencing the same issue, but I would really like to use concurrent read/write operations and lock only when necessary. For some odd reason, I only noticed this issue with DTLS, never with TLS. mbedtls_ssl_flush_output is called from mbedtls_ssl_read() only for DTLS, this could explain it.  Did you solve the issue nicely on your end, or did you end up locking concurrent read/write operations?

**Comment by awakecoding** [comment]
@pjbakker is there a particular reason why there is no lock for read and write operations, such that a read operation would acquire the write lock if it needs to write, and prevent such race conditions? This might be enough to make concurrent read/write operations possible, since the vast majority of the execution time, there is no issue with doing them parallel.

**Comment by pjbakker** [comment]
@awakecoding None other than that the premise of the design of most mbed TLS contexts is that they are only used within a single thread and thus do not need such locking.

**Comment by awakecoding** [comment]
@pjbakker I have started making some changes on my own branch (https://github.com/wayk/mbedtls/tree/wayk). I'm trying to keep the changes clean enough for inclusion later on. mbedtls_ssl_context has a few logical group of member variables that could be locked independently. In my case, I added a lock to protect against a race condition I encountered around the "Record layer (outgoing data)" section (the out_* variables), and it appears to be enough. In theory, I would also need to lock session renegotiation as well, but it is currently disabled in my code. One other thing I did was to extended the mutex API to add support for trylock. mbedtls_ssl_read checks if there is pending data to be flushed, it is better to attempt locking and not bother if we can't lock: if it is already locked, some other thread is already flushing the out data anyway, so we're good. My development is not completed yet, so I can't know if the changes I made are sufficient, but so far I no longer encounter the race conditions I used to have.
